[fuse] 뉴스스탠드 #1 PR

silvertae opened the pull request:
![스크린샷 2023-05-26 오전 11 59 09](https://github.com/codesquad-members-2023/fe-max--newsstand/assets/76121068/ed9f9239-12aa-4b29-b89c-f6d47bf0bcd1)

리뷰 요청이 늦어져서 정말 죄송합니다.
구현은 많이 못 하더라도 노션이나 리드미를 좀 잘 정리해서 리뷰 요청드려야 했었는데 정리가 너무 안 되어있네요.

## 현재까지의 시도
- scss 7-1 pattern 적용해서 style 적용하기
- 기존 html로 구현되어 있던 부분들을 클래스를 활용해서 다시 구현하기
- template을 이용하지 않고 createElement(), append() 사용해보기

https://react.dev/learn/thinking-in-react 에 나와있는 설명을 참고해서 리액트와 비슷한 느낌으로 컴포넌트들을 구현하고 싶었는데, typescript도 잘 사용하지 못하는 상태에서 시도하려니 자꾸 타입 에러가 나네요. 😅

우선 상태에 따른 변화는 생각하지 않고 정적인 부분들만 구현해보고 있는 정도입니다.
리액트의 하위 호환 느낌을 바닐라로 구현하려면 state, props에 대한 이해와 이것을 어떻게 구성하고 넘겨줄지에 대해서 고민이 더 필요할 것 같습니다. Flux 패턴과 같은 부분은 나중에 천천히 적용해봐야 될 것 같습니다.

## 작업 현황
- 네이버 뉴스데이터 크롤링 -> 현재 원래 사용했던 크롤링 방식이 오류가 나서 기존에 만들어 두었던 json 데이터를 사용하고 있습니다.
- 프로그램 설계
- html, scss 레이아웃 작업
- 기존 설계했던 컴포넌트 계층 나누기

노션 링크
https://seasoned-fire-6f4.notion.site/1-2a20f362fac04192ba0353d9846bc40a

silvertae committed:
chore: 타입스크립트 초기 개발환경 세팅

- typescript 설치
- eslint, prettier 설치 및 설정

silvertae committed:
chore: vite의 vanilla-ts template 적용

- npm create vite를 이용해서 개발 환경 설정

silvertae committed:
chore: scss 세팅 및 폴더 구조 변경

- sass를 devDependencies에 추가
- scss로 스타일 적용
- src 폴더 구조 변경 및 파일 이동

silvertae committed:
docs: 노션 링크 리드미에 추가

saejinpark commented on index.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
```
ko로 바꿔주세요 여기는 한국입니다.

namse replied:
젶토, "바꿔주세요" 같이 강한 주장을 하려면 합리적인 이유를 다셔야해요. '여기는 한국이다' 정도로는 강한 주장을 할만한 이유가 아니에요.

namse replied:
`저는 여기에 보통 ko를 씁니다. ko가 아니라 여기 en 쓰신 이유가 궁금합니다. 저도 en을 써야하는건지 궁금하네요.`

saejinpark commented on .eslintrc.json:
```diff
@@ -0,0 +1,13 @@
+    "sourceType": "module"
+  },
+  "plugins": ["@typescript-eslint"]
+}
```
나중에 eslint 쓰는거 알려주세요.. 히히

silvertae committed:
chore: 레이아웃 작업 전 기본 세팅

- 불필요한 Vite 템플릿 제거
- CSS minireset 적용
- 폰트 cdn 링크 연결
- 파비콘 추가

silvertae committed:
chore: 아이콘 svg 파일 assets에 추가

silvertae committed:
feat: 크롤링 코드와 획득한 JSON 데이터 추가

- crawling 폴더에 크롤링할 때 사용한 js 코드들 추가
- 크롤링한 데이터로 만들어 둔 json 데이터 파일 추가

silvertae committed:
style: 레이아웃 작업 전 폰트 관련 설정

- 웹폰트를 다운로드 받는데 걸리는 지연 시간을 줄이기 위해 기존 Pretendard
기본 폰트가 아닌 용량이 적은 가변 다이나믹 서브셋 폰트를 받아오도록 변경
- typography 관련 scss 파일 추가

silvertae committed:
style: scss 파일 구조 변경 및 변수 선언

- @use, @forward 문법 활용해서 scss 파일 구조 변경
- scss의 기능들을 적극 활용하기 위해서 기존 :root에서 선언한 css 변수들 삭제
- abstracts 폴더에 mixins, variables 파일을 만들어서 폰트와 컬러 관련 설정

silvertae committed:
style: 헤더 부분 레이아웃 작업

- html로 레이아웃 잡기
- _header.scss로 헤더 부분 스타일 적용

silvertae committed:
style: 자동 롤링 영역 레이아웃 작업

- _newsbar.scss 파일에 스타일 추가
- 우선 후에 애니메이션 기능 고려하지 않고 레이아웃 작업만 진행
- 후에 더 좋은 클래스명이 생각나면 수정

silvertae committed:
Merge branch 'layout' into W1

silvertae committed:
style: 메인 영역의 헤더 레이아웃 작업

- 메인에서 탭 버튼들이 있는 헤더부분 레이아웃 작업
- 뉴스바 기사 호버시 밑줄 추가
- 메인 콘텐츠 영역 그리드 스타일 작업 진행 예정

silvertae committed:
style: 메인 콘텐츠 그리드 뷰 레이아웃 완성

- CSS grid로 그리드 뷰에서의 레이아웃 완성
- 언론사 로고 src는 크롤링한 데이터로 넣어줄 예정

silvertae committed:
Merge branch 'layout' into W1

silvertae committed:
style: 언론사 로고 이미지 추가와 그리드뷰 이동 버튼 추가

- 언론사 로고 이미지 추가
- 그리드 뷰 양쪽에 화살표 버튼 추가와 배치
- position 관련 mixin들 추가

silvertae committed:
feat: UI를 컴포넌트 단위로 계층 구조 나누기 (header, roller)

- html로 구현했던 레이아웃을 컴포넌트 단위로 쪼개서 만들고 있다.
- 우선 main.ts에서 작업하고 후에 분리 예정
- 현재 header, rollingSection 작업 완료, 후에 main 영역 작업 예정
- 지금은 state, props 신경쓰지 말고 레이아웃만 완성하기

silvertae committed:
feat: 메인 영역 헤더와 GridView 컴포넌트로 렌더링하기

- 메인 영역을 headerTab과 grid와 list View가 들어갈 content 영역으로 나누어 구현
- 기존에 img 태그로 넣었던 아이콘들을 html 구조를 단순화 하기 위해 button 요소의
background-image로 변경
- 각 언론사 로고들이 들어갈 grid를 어떻게 구상할지 고민

silvertae committed:
feat: 그리드에서 각 언론사 로고 이미지 렌더링 기능 추가

- GridView 컴포넌트에서 render()를 호출해 state에 있는 정보들을 가지고
logoUrl을 가져와서 이를 img 요소의 src에 할당한다.
- state는 data 폴더에 넣어둔 json 데이터를 가져와서 사용하게 했다.
- 이제 컴포넌트를 파일별로 나누는 작업과 state를 어떻게 전달할지 작업해야한다.

silvertae committed:
feat: main.ts에 모여있던 컴포넌트들 파일로 분리

- 각 컴포넌트들을 파일로 분리해서 만들었다.
- 전역에 선언해두고 사용했던 state는 일단 각 클래스의 인스턴스를 만들때
상위 요소가 하위요소에 props 인자로 전달하는 방식으로 했다.
- props의 type에 오류가 나고 있어서 각 타입들을 선언해야 한다.

altmit commented on src/components/GridView.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  render(props) {
+    this.gridItems.forEach((item, idx) => {
+      const itemImg = item.children[0] as HTMLImageElement;
+      itemImg.src = props.newsData[idx].logoURL;
+      itemImg.alt = props.newsData[idx].name;
+    });
```
데이터가 24개가 확신된 코드인 거 같습니다.

구독중인 언론사만 볼 경우 몇개의 newsData가 올 지 모르기 때문에 24개 미만의 경우에도 넘어온 데이터의 개수 만큼 추가하는 로직이면 재사용 가능한 함수가 될 것 같습니다.

silvertae replied:
구독중인 언론사를 보여주는 부분에 대해서는 미처 설계를 못 했네요. 말씀하신 것처럼 필요한 데이터의 개수에 맞게 이미지를 가져올 수 있도록 재사용 가능하도록 만들어보겠습니다.

hjsong333 commented on src/main.ts:
```diff
@@ -0,0 +1,19 @@
+
+function App() {
+  const app = document.querySelector('#app');
+  app?.append(newsStand.element);
```
옵셔널 체이닝 간편하고 좋지만 가독성은 if문이 더 좋은 것 같아요.
또, if문을 사용해서 처리하면 추가적인 처리가 필요할 때 더 좋을 것 같습니다.

silvertae replied:
vscode가 자동으로 저렇게 처리해줘서 일단 급하게 넘어가긴 했는데 확실히 if문이 더 가독성이 좋을 것 같네요. 감사합니다.

hjsong333 commented on src/components/Main.ts:
```diff
@@ -0,0 +1,54 @@
+  constructor(props) {
+    this.element = document.createElement('main');
+
+    {
+      const headerTab = document.createElement('header');
+      headerTab.classList.add('header--main');
+
+      {
+        const tabBtnMenu = document.createElement('div');
+        tabBtnMenu.classList.add('tab-btn__menu');
+
+        const showAllBtn = document.createElement('div');
+        showAllBtn.classList.add('tab-btn__item', 'active');
+        showAllBtn.textContent = '전체 언론사';
+
+        const showSubscribedBtn = document.createElement('div');
+        showSubscribedBtn.classList.add('tab-btn__item');
+        showSubscribedBtn.textContent = '내가 구독한 언론사';
+
+        tabBtnMenu.append(showAllBtn, showSubscribedBtn);
+        headerTab.append(tabBtnMenu);
+      }
+
+      {
+        const viewBtnMenu = document.createElement('div');
+        viewBtnMenu.classList.add('view-btn__menu');
+
+        const listBtn = document.createElement('button');
+        listBtn.classList.add('view-btn__item', 'list-icon');
+
+        const gridBtn = document.createElement('button');
+        gridBtn.classList.add('view-btn__item', 'grid-icon', 'active');
+
+        viewBtnMenu.append(listBtn, gridBtn);
+        headerTab.append(viewBtnMenu);
+      }
+
+      this.element.append(headerTab);
+    }
```
코드 블록이 특별한 역할을 하는 것 같지는 않은데 구분을 위해서 해놓으신 걸까요?

silvertae replied:
특별한 역할은 없지만 일단 메인영역에서 헤더 부분과 콘텐츠가 표시되는 부분을 구분해서 보려고 나눠두었습니다.

jsh3418 commented on src/components/GridView.ts:
```diff
@@ -0,0 +1,44 @@
+    this.element = document.createElement('div');
+    this.element.classList.add('grid-container');
+
+    const leftBtn = document.createElement('button');
+    this.leftBtn = leftBtn;
+    leftBtn.classList.add('btn', 'btn--left');
+
+    const rightBtn = document.createElement('button');
+    this.rightBtn = rightBtn;
+    rightBtn.classList.add('btn', 'btn--right');
```
따로 변수에 할당하신 이유가 있으신가요? 이렇게 하면 될 것 같아서요.
```ts
this.leftBtn = document.createElement('button');
this.leftBtn.classList.add('btn', 'btn--left');
```

jsh3418 commented on src/scss/layout/_main.scss:
```diff
@@ -0,0 +1,94 @@
+      @include pos-center();
+    }
+  }
+}
```
depth가 깊어질수록 가독성이 떨어지고 유지보수할 때 힘들겠더라구요. depth를 줄이는 방법을 고민해보시면 좋을 것 같아요.

silvertae committed:
chore: tsconfig 설정 추가
